### DISCUSSION TITLE
Best practices for rapid interactive development in Lua

### DISCUSSION THREAD

**Discussion by mattwarkentin** [quarto-dev/quarto-cli#4039] [Q&A]
Hi Quarto team,

Is it possible to step into an interactive Lua interpreter with access to the pandoc Lua API? I find it particularly difficult to write in Lua without being able to test the Pandoc and Quarto Lua APIs interactively without actually having to run/render a document.

How do you all go about writing Lua code? Any suggestions of best practices for rapid interactive iteration? I would love to be able to access the Quarto Lua API also, but I understand that this might not be possible outside of the context of document rendering.

I do think that if Quarto could somehow support a way to step into a Lua interpreter with access to the Pandoc and Quarto APIs it could really improve development. Open to any and all suggestions.

**Comment by cscheid** [comment]
It's unfortunately not possible with the capabilities in current Pandoc versions (and trust me, it hurts as much for us as it does for you!) Pandoc 3 improves things somewhat with `pandoc lua`, a command that runs the Lua interpreter from pandoc natively. Note that it does not currently include REPL support.

A few things I do:

- Use test documents with formats that don't take long to render (specifically, the LaTeX toolchain is slow)
- Take advantage specifically of `format: native`, which prints a representation of the Pandoc AST rather than any specific format
- Write small self-contained functions that you can test with `quarto run foo.lua`. This is not an interactive shell, but it's a way to run a pure Lua module through an interpreter with pandoc and quarto APIs

Improving our Lua DX is definitely on our radar! We appreciate hearing about your experiences with it.

**Reply by mattwarkentin** [reply]
Thanks for the detailed response. This is quite helpful.
